# 02. Pandas

Pandas란, 엑셀시트 같은 데이터 조작 및 분석 도구

---

## 주요 특징
- 넘파이를 기반으로 한다
- 1차원 배열 형태의 데이터인 Series, 2차원 표 구조의 데이터인 DataFrame
- 강력한 전처리
- Panel Data (Pandas라는 이름의 어원)

---

## 기본 세팅

In [1]:
import numpy as np
import pandas as pd
print(pd.__version__)

3.0.5


---

## Series와 DataFrame

- **Series**: 1차원. 엑셀의 **열 하나**
- **DataFrame**: 2차원 표. **Series 여러 개를 옆으로 붙인 것**

df["Name"]  -> Series  (열 하나를 꺼냄)  
df[["Name"]] -> DataFrame (열 목록을 꺼냄, 결과가 표)

- pd.Series(데이터, index, name): 데이터를 시리즈로 변환
- pd.DataFrame(데이터, index, columns): 데이터를 데이터프레임으로 변환

In [2]:
data = {
    'Name':['Jackson', 'Emma', 'Noah', 'James'],
    'Weight':[68, 74, 77, 78],
    'Option':[True, True, False, False],
    'Rate':[0.21, 1.1, 0.89, 0.91],
    'Type':['A','A','C','C']
}

| Name | Weight | Option | Rate | Type |
|---|---|---|---|---|
| Jackson | 68 | True | 0.21 | A |
| Emma | 74 | True | 1.10 | A |
| Noah | 77 | False | 0.89 | C |
| James | 78 | False | 0.91 | C |


### 데이터프레임으로 변환
- pd.DataFrame(데이터)

In [3]:
df = pd.DataFrame(data)
df

,Name,Weight,Option,Rate,Type
0,Jackson,68,True,0.21,A
1,Emma,74,True,1.10,A
2,Noah,77,False,0.89,C
3,James,78,False,0.91,C


In [4]:
print("df의 타입: ", type(df))

df의 타입:  <class 'pandas.DataFrame'>


### DataFrame -> Numpy 배열

- 데이터프레임.to_numpy()  

컬럼 이름과 인덱스가 사라지고 값만 남는다

In [5]:
df_converted = df.to_numpy()
df_converted

array([['Jackson', 68, True, 0.21, 'A'],
       ['Emma', 74, True, 1.1, 'A'],
       ['Noah', 77, False, 0.89, 'C'],
       ['James', 78, False, 0.91, 'C']], dtype=object)

In [6]:
print("df_converted.dtype: ", df_converted.dtype)

df_converted.dtype:  object


---

## 데이터 분석

- 데이터 첫 5행  
df.head()


- 데이터 마지막 5행  
df.tail()


- 데이터프레임의 크기  
df.shape


- 열 하나 -> Series  
df["Name"]


- 열 목록 -> DataFrame  
df[["Name"]]  
(안쪽 [ ]는 "뽑을 열 이름들의 리스트"이다)

---

## titanic.csv 데이터 분석

> 데이터: Kaggle Titanic 대회의 train.csv를 titanic.csv로 저장해 사용  
> .gitignore(*.csv)로 제외되어 레포에는 포함되지 않음  
> 승객 891명, 12개 컬럼

### 파일에서 읽기
- read_csv(csv파일)

In [7]:
df = pd.read_csv("titanic.csv")
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


### 데이터 분석 (head(), info(), describe() 등등)

#### 1. head()

In [8]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


- 출력에 NaN이 보인다 = 결측치(값이 비어 있음)
- NaN은 "Not a Number". 숫자 컬럼이든 문자 컬럼이든 빈칸은 NaN으로 표시된다

#### 2. info()

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


### info() 읽는 법

| 항목 | 의미 |
|---|---|
| RangeIndex: 891 entries | 전체 891행 |
| Non-Null Count | 값이 있는 행의 개수 |
| Dtype | 자료형 (object = 문자열) |

**전체 891과 다른 컬럼이 결측치가 있는 컬럼이다.**
- Age: 714 -> 177개 비어 있음
- Cabin: 204 -> 687개 비어 있음 (77%)
- Embarked: 889 -> 2개 비어 있음

#### 3. describe()

In [10]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### describe() 읽는 법

- 숫자 컬럼만 나온다 (Name, Sex 같은 문자 컬럼은 제외)
- 25% / 50% / 75% = **사분위수**. 값을 작은 순으로 줄 세웠을 때 1/4, 2/4, 3/4 지점
  - Age 25% = 20.1 -> 승객의 1/4이 20세 이하
  - 50%는 중앙값(median)

### 여기서 알 수 있는 것

1. **count가 컬럼마다 다르다** — Age만 714. 결측치가 여기서도 드러난다
2. **Fare는 평균 32.20, 중앙값 14.45, 최대 512.33**
   → 평균이 중앙값의 2배. 소수의 비싼 티켓이 평균을 끌어올렸다
   → 이런 컬럼은 "평균 요금 32달러"라고 말하면 실제와 동떨어진다
3. **Survived의 평균 0.3838 = 생존율 38.4%**
   0/1 값만 있는 컬럼은 평균이 곧 비율이다

#### 4. describe().T

In [11]:
# 전치
df.describe().T

,count,mean,std,min,25%,50%,75%,max
PassengerId,891.0,446.000000,257.353842,1.00,223.5000,446.0000,668.5,891.0000
Survived,891.0,0.383838,0.486592,0.00,0.0000,0.0000,1.0,1.0000
Pclass,891.0,2.308642,0.836071,1.00,2.0000,3.0000,3.0,3.0000
Age,714.0,29.699118,14.526497,0.42,20.1250,28.0000,38.0,80.0000
SibSp,891.0,0.523008,1.102743,0.00,0.0000,0.0000,1.0,8.0000
Parch,891.0,0.381594,0.806057,0.00,0.0000,0.0000,0.0,6.0000
Fare,891.0,32.204208,49.693429,0.00,7.9104,14.4542,31.0,512.3292


컬럼이 많을 때 가로로 길어져서 화면에 다 안 들어오기 때문에 행과 열을 뒤집는 것

In [12]:
# 특정 컬럼의 데이터들만 뽑아서 보기
df[["Age", "Name"]]

,Age,Name
0,22.0,"Braund, Mr. Owen Harris"
1,38.0,"Cumings, Mrs. John Bradley (Florence Briggs Th..."
2,26.0,"Heikkinen, Miss Laina"
3,35.0,"Futrelle, Mrs. Jacques Heath (Lily May Peel)"
4,35.0,"Allen, Mr. William Henry"
...,...,...
886,27.0,"Montvila, Rev. Juozas"
887,19.0,"Graham, Miss Margaret Edith"
888,NaN,"Johnston, Miss Catherine Helen ""Carrie"""
889,26.0,"Behr, Mr. Karl Howell"


## loc / iloc

| | 기준 | 예시 |
|---|---|---|
| loc | **레이블(이름)** | df.loc[0, "Age"] |
| iloc | **위치(정수 순서)** | df.iloc[0, 5] |

타이타닉은 인덱스가 0,1,2...라 둘이 같아 보이지만,
**행을 걸러내는 순간 달라진다.**

슬라이싱도 다르다
- df.loc[0:3]  -> 0,1,2,3  (끝 포함)
- df.iloc[0:3] -> 0,1,2    (끝 제외, 파이썬 기본과 동일)

## 조건 필터링

- `and`가 아니라 `&`를 쓴다
  df["Age"]>50 의 결과는 True/False가 891개 든 Series다.
  `and`는 값 하나끼리만 비교 가능 -> 에러.
  `&`는 같은 자리끼리 비교한다 (Numpy 논리 인덱싱과 같은 원리)

- **괄호 필수**
  `&`가 `>`보다 먼저 계산되기 때문
  df["Age"] > 50 & df["Age"] < 60      ❌
  (df["Age"] > 50) & (df["Age"] < 60)  ⭕

- or는 `|`, 부정은 `~`

In [13]:
print(df.loc[0:3].shape)    # (4, ...) 끝 포함
print(df.iloc[0:3].shape)   # (3, ...) 끝 제외

(4, 12)
(3, 12)


In [14]:
df.sort_values("Fare", ascending=False).head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
679,680,1,1,"Cardeza, Mr. Thomas Drake Martinez",male,36.0,0,1,PC 17755,512.3292,B51 B53 B55,C
258,259,1,1,"Ward, Miss Anna",female,35.0,0,0,PC 17755,512.3292,NaN,C
737,738,1,1,"Lesurer, Mr. Gustave J",male,35.0,0,0,PC 17755,512.3292,B101,C
88,89,1,1,"Fortune, Miss Mabel Helen",female,23.0,3,2,19950,263.0000,C23 C25 C27,S
438,439,0,1,"Fortune, Mr. Mark",male,64.0,1,4,19950,263.0000,C23 C25 C27,S


In [15]:
# Age 컬럼의 값이 50초과 60미만인 행만 뽑아서 보기
tmp_df = df.loc[(df["Age"]>50) & (df["Age"]<60)]
print(tmp_df.shape)
tmp_df.head()

(38, 12)


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
11,12,1,1,"Bonnell, Miss Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
15,16,1,2,"Hewlett, Mrs. (Mary D Kingcome)",female,55.0,0,0,248706,16.0000,NaN,S
94,95,0,3,"Coxon, Mr. Daniel",male,59.0,0,0,364500,7.2500,NaN,S
124,125,0,1,"White, Mr. Percival Wayland",male,54.0,0,1,35281,77.2875,D26,S


In [16]:
# 걸러낸 뒤에는 인덱스가 띄엄띄엄해진다
print(tmp_df.index[:5])

print(tmp_df.iloc[0]["Name"])   # 위에서 첫 번째 행
# print(tmp_df.loc[0])          # KeyError — 0번은 걸러져서 없다

Index([6, 11, 15, 94, 124], dtype='int64')
McCarthy, Mr. Timothy J


### 파생 변수 만들기

SibSp(형제/배우자 수) + Parch(부모/자녀 수) = 같이 탄 가족 수

원래 데이터에 없던 컬럼을 계산해서 추가한 것.
두 컬럼을 따로 보는 것보다 "가족과 함께 탔는가"가
해석하기 쉬운 정보다.

df["새컬럼"] = 계산식  -> 열이 추가된다

In [17]:
# 새로운 열 추가
df["family"] = df["SibSp"]+df["Parch"]

In [18]:
# 행 삭제: axis 기본값이 0이라 생략 가능
# df.drop(0) / df.drop([1, 2, 3])

# 열 삭제: columns= 를 쓰면 axis를 외울 필요가 없고 코드도 읽기 쉽다
df = df.drop(columns=["SibSp", "Parch"])

df["Age"].max(), df["Age"].min(), df["Age"].mean(), df["Age"].median()

In [19]:
# 범주형 컬럼은 value_counts로 본다
print(df["Sex"].value_counts())
print(df["Embarked"].unique())        # 고유값 목록
print(df["Ticket"].nunique())         # 고유값 개수

# normalize=True -> 개수 대신 비율
df["Survived"].value_counts(normalize=True)

Sex
male      577
female    314
Name: count, dtype: int64
<StringArray>
['S', 'C', 'Q', nan]
Length: 4, dtype: str
681


Survived
0    0.616162
1    0.383838
Name: proportion, dtype: float64

## 그룹화 (groupby)

원리: 1) 쪼개기 → 2) 적용 → 3) 결합

df.groupby("Sex")["Age"].mean() 이 하는 일

1. **쪼개기**: 891행을 Sex 값에 따라 male 577행 / female 314행으로 나눈다
2. **적용**: 각 덩어리에서 Age의 평균을 따로 구한다
3. **결합**: 두 결과를 하나의 Series로 합친다

`df.groupby(기준컬럼)[대상컬럼].집계함수()`

In [20]:
# 성별에 따른 'Age'의 평균 구하기
df.groupby("Sex")['Age'].mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

In [21]:
# 성별마다의 'Pclass'별 'Age'의 평균 구하기
df.groupby(["Sex", "Pclass"])["Age"].mean()

Sex     Pclass
female  1         34.611765
        2         28.722973
        3         21.750000
male    1         41.281386
        2         30.740707
        3         26.507589
Name: Age, dtype: float64

In [22]:
# 여러 통계를 한 번에
df.groupby("Sex")["Age"].agg(["mean", "median", "count"])

,mean,median,count
Sex,,,
female,27.915709,27.0,261
male,30.726645,29.0,453


**count를 같이 보는 이유**

Sex별 인원은 male 577 / female 314인데
count는 453 / 261로 나온다. **결측치를 빼고 센 것**이다.
평균이 몇 명 기준인지 항상 같이 확인해야 한다.

In [23]:
# groupby 결과는 인덱스가 2단으로 쌓인 형태
# reset_index()로 평범한 표로 되돌린다
df.groupby(["Sex", "Pclass"])["Age"].mean().reset_index()

,Sex,Pclass,Age
0,female,1,34.611765
1,female,2,28.722973
2,female,3,21.750000
3,male,1,41.281386
4,male,2,30.740707
5,male,3,26.507589


In [24]:
# 전체 컬럼 결측치 한 번에
print(df.isna().sum())

# 비율로 보기
print(df.isna().mean().sort_values(ascending=False))

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
Ticket           0
Fare             0
Cabin          687
Embarked         2
family           0
dtype: int64
Cabin          0.771044
Age            0.198653
Embarked       0.002245
Pclass         0.000000
Survived       0.000000
PassengerId    0.000000
Name           0.000000
Ticket         0.000000
Sex            0.000000
Fare           0.000000
family         0.000000
dtype: float64


In [25]:
# 확인: fillna는 원본을 바꾸지 않는다
print("처리 전 평균 :", df["Age"].mean())
_ = df.fillna(0)
print("fillna 후 평균:", df["Age"].mean())   # 그대로다

처리 전 평균 : 29.69911764705882
fillna 후 평균: 29.69911764705882


In [26]:
# 0으로 일괄 처리하면 어떻게 되는지 확인
print("원본 평균      :", df["Age"].mean())
print("0으로 채운 평균:", df["Age"].fillna(0).mean())

원본 평균      : 29.69911764705882
0으로 채운 평균: 23.79929292929293


In [27]:
# 컬럼 성격에 맞게 처리한다

# Age: 숫자 -> 중앙값
# 평균이 아니라 중앙값을 쓰는 이유: 극단값에 덜 흔들린다
df["Age"] = df["Age"].fillna(df["Age"].median())

# Embarked: 범주형 -> 최빈값
# mode()는 동점일 수 있어 Series를 반환하므로 [0]으로 첫 값을 꺼낸다
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])

# Cabin: 687/891 = 77% 결측 -> 채울 근거가 없으니 컬럼 삭제
df = df.drop(columns=["Cabin"])

print(df.isna().sum())

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
Ticket         0
Fare           0
Embarked       0
family         0
dtype: int64


### 결측치 처리 기준

| 상황 | 처리 |
|---|---|
| 숫자 컬럼 | 중앙값(median). 평균은 극단값에 흔들린다 |
| 범주형 컬럼 | 최빈값(mode) |
| 결측이 너무 많음(대략 50% 이상) | 컬럼 삭제 |
| 행이 몇 개 안 됨 | dropna()로 행 삭제 |

**inplace=True는 쓰지 않는다.** pandas 3.x에서 동작이 바뀌었고,
`df = df.어쩌구()` 재대입이 어디서 뭐가 바뀌었는지 코드에 드러난다.

In [28]:
# 나이를 구간으로 묶기 (파생 변수)
df["AgeGroup"] = pd.cut(
    df["Age"],
    bins=[0, 12, 20, 40, 60, 100],
    labels=["아동", "청소년", "청년", "중년", "노년"]
)

# 나이대별 생존율
df.groupby("AgeGroup", observed=True)["Survived"].agg(["mean", "count"])

,mean,count
AgeGroup,,
아동,0.579710,69
청소년,0.381818,110
청년,0.364769,562
중년,0.390625,128
노년,0.227273,22


### 주의: 이 결과의 한계

Age 결측치 177개를 중앙값 28로 채웠다.
28은 (20, 40] 구간이므로 **177명이 전부 "청년"에 들어갔다.**
청년 562명 중 실제 나이를 아는 사람은 385명뿐이다.

즉 청년의 생존율 0.365는 만들어낸 값이 섞인 숫자다.
결측치를 채우는 처리는 이렇게 **뒤쪽 분석 결과까지 오염시킨다.**

대안: 결측치를 채우지 않고 pd.cut에 그대로 넣으면
NaN은 어느 구간에도 안 들어가서 "모름"으로 분리된다.